File: 3_Triangle_Algorithm/tri_algo.m
# Computational Linear Algebra Ep#3 Uppper Triangle Algorithm in Octave
# By: Nick Space Cowboy

clear all;
close all;

function x = solve_utri(A,b) # creates the function solve_utri
	n = length(b);	# To help with indexing
	x = zeros(n,1);	# creates a x vector to be filled
	for k=n:-1:2	# starts iterating through matrix 
	x(k) = b(k)/A(k,k);
	b(1:k-1) = b(1:k-1) - x(k)*A(1:k-1,k);
	end
	x(1) = b(1)/A(1,1);
endfunction

b = randi(100, 10, 1) # generates random 10d b vector
A = triu(randi(100, 10, 10),0)# genertates random 10x10 A matrix
x = solve_utri(A,b)	# solves the system with our created function

check = A*x	# checks that our result is in fact correct

File: 3_Triangle_Algorithm/tri_algo.py
# Computational Linear Algebra Ep#3 Uppper Triangle Algorithm
# By: Nick Space Cowboy

import numpy as np

class Cowboy_Lin_Alg(object):
	def solve_utri(self, Utri, b):
		n = len(Utri) # row dimension of the Utri matrix
		x = np.zeros_like(b, dtype=float)
		for i in range(n - 1, -1, -1):	# loop to iterate through row index
			x[i] += b[i] / Utri[i,i]
			for j in range(n-1, i, -1):	# loop to iterate through the off diagonal Sum part
				x[i] += (- (Utri[i, j] * x[j])) / Utri[i,i]
		return x
		
			
if __name__ == "__main__":
	cla = Cowboy_Lin_Alg()
	A = np.array(np.random.randint (0,100,(4,4)))
	U = np.triu(A)
	print("U: " + str(U))
	b = np.array(np.random.randint(0, 100, (4,1)))
	print("b:" + str(b))
	x = cla.solve_utri(U,b)
	print("x: " + str(x))
	print("Check: " + str(U.dot(x)))
	
